plugins/vbe/fu-vbe-simple-device.c

```diff
@@ -0,0 +1,901 @@
+ * Select the best configuration for the model, then write that configuration
+ * to the device.
+ *
+ * @fit: FIT to write
+ * @dev: Device to use
+ * @compat_list: List of compatible properties for this model, if any
+ * @progress: Progress information to update (this is simplistic and assumes
+ * that each image takes the same time to write)
+ * @error: Returns error information if this function returns FALSE
+ * Returns: TRUE on success, FALSE on failure
+ */
+static gboolean
+process_fit(struct fit_info *fit,
+	    struct _FuVbeSimpleDevice *dev,
+	    GList *compat_list,
+	    FuProgress *progress,
+	    GError **error)
+{
+	struct vbe_simple_state *state = &dev->state;
+	gint best_prio = INT_MAX;
+	const gchar *cfg_name;
+	const gchar *version;
+	const GList *entry;
+	gint cfg_count = 0;
+	gint best_cfg = 0;
+	gint cfg;
+
+	g_debug("model: ");
+	for (entry = g_list_first(compat_list); entry; entry = g_list_next(entry))
+		g_debug("   %s", (gchar *)entry->data);
+
+	for (cfg = fit_first_cfg(fit); cfg > 0; cfg_count++, cfg = fit_next_cfg(fit, cfg)) {
+		gint prio = check_config_match(fit, cfg, compat_list);
+		g_debug("config '%s': priority=%d", fit_cfg_name(fit, cfg), prio);
+		if (prio >= 0 && (!best_cfg || prio < best_prio)) {
+			best_cfg = cfg;
+			best_prio = prio;
+		}
+	}
+
+	g_debug("cfg_count=%d, best_cfg=%d", cfg_count, best_cfg);
+	if (!cfg_count) {
+		g_set_error(error, FWUPD_ERROR, FWUPD_ERROR_READ, "FIT has no configurations");
+		return FALSE;
+	}
+
+	if (!best_cfg) {
+		g_set_error(error, FWUPD_ERROR, FWUPD_ERROR_READ, "No matching configuration");
+		return FALSE;
+	}
+	cfg_name = fit_cfg_name(fit, best_cfg);
+	version = fit_cfg_version(fit, best_cfg);
+	if (!version) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_READ,
+			    "Configuration '%s' has no version",
+			    cfg_name);
+		return FALSE;
+	}
+
+	g_debug("Best configuration: '%s', priority %d, version %s", cfg_name, best_prio, version);
+
+	if (!process_config(fit, best_cfg, dev, progress, error))
+		return FALSE;
+
+	g_free(state->cur_version);
+	g_free(state->status);
+	state->finish_time = time(NULL);
+	state->cur_version = g_strdup(version);
+	state->status = g_strdup("completed");
+
+	return TRUE;
+}
+
+static gboolean
+fu_vbe_simple_device_write_firmware(FuDevice *self,
+				    FuFirmware *firmware,
+				    FuProgress *progress,
+				    FwupdInstallFlags flags,
+				    GError **error)
+{
+	struct _FuVbeSimpleDevice *dev = FU_VBE_SIMPLE_DEVICE(self);
+	g_autoptr(GBytes) bytes = NULL;
+	struct fit_info fit;
+	GList *compat_list;
+	FuVbeDevice *vdev;
+	const guint8 *buf;
+	gsize size = 0;
+	gint ret;
+
+	vdev = FU_VBE_DEVICE(self);
+	g_return_val_if_fail(FU_IS_VBE_DEVICE(self), FALSE);
+
+	compat_list = fu_vbe_device_get_compat_list(vdev);
+
+	bytes = fu_firmware_get_bytes(firmware, error);
+	if (!bytes)
+		return FALSE;
+	buf = g_bytes_get_data(bytes, &size);
+
+	g_debug("Size of FIT: %#zx\n", size);
+	ret = fit_open(&fit, buf, size);
```

Comment:
It's a bit odd you don't pass in the error here to `fit_open`.  I guess that's because you're trying to keep glib out of `fit.c` still?

Reply:
Yes that's right. I want the fit library to use usable by non-glib code, since it might be used in firmware, or other projects which use libfdt.